File: java/pascals-triangle.java
/*
Given an integer numRows, return the first numRows of Pascal's triangle.
In Pascal's triangle, each number is the sum of the two numbers directly above it as shown:
*/

class Solution {
    public List<List<Integer>> generate(int numRows) {
        List<List<Integer>> sol = new ArrayList<>();
        List<Integer> firstRow = new ArrayList<>();
        firstRow.add(1);
        sol.add(firstRow);
        for(int i = 1; i < numRows; i++) {
            List<Integer> prevRow = sol.get(i - 1);
            List<Integer> newRow = new ArrayList<>();
            newRow.add(1);

            for(int j = 1; j < i; j++) {
                newRow.add(prevRow.get(j - 1) + prevRow.get(j));
            }
            newRow.add(1);
            sol.add(newRow);
        }
        return sol;
    }
}

File: python/pascals-triangle.py
class Solution(object):
    def generate(self, numRows):
        """
        :type numRows: int
        :rtype: List[List[int]]
        """
        i = 1
        rows = []
        for i in range(0, numRows):
            if i == 0:
                rows.append([1])
            elif i == 1:
                rows.append([1, 1])
            else:
                new_row = [1]
                prev_row = rows[i - 1]
                for j in range(len(prev_row) - 1):
                    new_num = prev_row[j] + prev_row[j + 1]
                    new_row.append(new_num)
                new_row.append(1)
                rows.append(new_row)
        return rows




        
        
